File: 01_GOF23/DP_examples/Lua/03_Behavioral/Command/Command.lua
-- 命令模式的 Lua 实现

-- 命令接口
local ICommand = {
    execute = function(self)
        error("Subclass must implement execute method")
    end
}

-- 创建命令对象的元表
function ICommand:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

-- 具体命令实现
local ConcreteCommand = ICommand:new()

function ConcreteCommand:new(receiver)
    local o = ICommand:new()
    setmetatable(o, self)
    self.__index = self
    o.receiver = receiver
    return o
end

function ConcreteCommand:execute()
    if self.receiver then
        self.receiver:action()
    end
end

-- 导出模块
return {
    ICommand = ICommand,
    ConcreteCommand = ConcreteCommand
}

File: 01_GOF23/DP_examples/Rust/03_Behavioral/Command/command.rs
// 命令实现

// 导入接收者接口
use crate::receiver::IReceiver;
use std::sync::Arc;

// 命令接口
pub trait ICommand {
    fn execute(&self);
}

// 具体命令实现
pub struct ConcreteCommand {
    receiver: Box<dyn IReceiver>,
}

impl ConcreteCommand {
    pub fn new(receiver: Box<dyn IReceiver>) -> Self {
        ConcreteCommand {
            receiver,
        }
    }
}

impl ICommand for ConcreteCommand {
    fn execute(&self) {
        self.receiver.action();
    }
}

// 类型别名，使用Arc管理命令对象
pub type CommandPtr = Arc<dyn ICommand>;
